File: rust/project/constraints.rb
module RustProjection
  module Projector
    module_function

    # `admits:`/`pattern:` (`attribute_collector.rb`'s own keyword args) sit
    # on the SAME `Attribute` shape wherever `attribute()` is called —
    # a value object's own field (`EmailAddress.address, pattern: ...`) or
    # an aggregate/command's own usage-level attribute (`ExternalTransfer`'s
    # `direction, MovementDirection, admits: "Account::LedgerDirection"`).
    # Ruby checks BOTH doors (`admission.rb`'s own `check_admitted`/
    # `check_patterns`, called from `Value.build`, and `admit_declared_set`,
    # called from `coerce_declared_arguments`) — this file is the shared
    # half both `types.rb` (the value-object door) and `commands.rb` (the
    # command/entity-command-argument door) call into.

    # The scalar `admitted_scalar`/`check_patterns` actually operate on —
    # bare already if the attribute's own declared type IS the scalar
    # (`String`), or the sole field of a single-field value object (the
    # "one-field holder" `admission.rb`'s own comment names). `nil` for
    # anything else (a multi-field VO, a closed-set VO with more than one
    # discriminant) — not a shape either door can unwrap either.
    def scalar_field_expr(value_expr, attr_type, value_objects_by_name)
      return value_expr if attr_type == "String"

      vo = value_objects_by_name[attr_type]
      return nil unless vo && vo[:attributes].size == 1

      "#{value_expr}.#{rust_ident_field(vo[:attributes].first[:name])}"
    end

    # An OPTIONAL attribute is `Option<T>` in the struct, not `T` — the two
    # callers below used to hand `scalar_field_expr` the bare field
    # (`self.description`) whether or not `attr[:optional]` held, which
    # compiled fine for every attribute nobody had yet paired `optional:`
    # with `admits:`/`pattern:` and broke the moment one was (a plain
    # `Option<String>` where `pattern::matches`'s own signature wants
    # `&str`). Ruby's own two doors (`admission.rb#admit_declared_set`,
    # `coercion.rb#check_patterns`) both skip the check outright on `nil`
    # — so this resolves a SCALAR EXPRESSION TO CHECK, only when the
    # attribute is required, or a REBINDING TARGET (the source `Option`
    # expression the caller must `if let Some(...)` around) when it's
    # not — never both, and the caller (`wrap_if_optional`, below) is
    # what actually decides which shape came back.
    def optional_scalar_expr(value_expr, attr, value_objects_by_name)
      return [scalar_field_expr(value_expr, attr[:type], value_objects_by_name), nil] unless attr[:optional]

      [scalar_field_expr(OPTIONAL_BINDING, attr[:type], value_objects_by_name), value_expr]
    end

    # `Some(OPTIONAL_BINDING) = &optional_source` — match ergonomics binds
    # OPTIONAL_BINDING as a reference into the Option's own contents
    # (`&String`/`&SomeVO`), exactly the shape `scalar_field_expr` already
    # produces unwrapped field access against (`.field` on a VO reference
    # derefs the same way `self.field` already did). A `None` optional
    # field skips the check entirely, matching Ruby exactly rather than
    # treating "not present" as "present and empty."
    OPTIONAL_BINDING = "__optional_value"

    def wrap_if_optional(check, optional_source)
      return check unless optional_source

      "if let Some(#{OPTIONAL_BINDING}) = &#{optional_source} { #{check} }"
    end

    # `admitted_members`/`chapter_of`, read directly (`admission.rb`):
    # `"Aggregate::SetName"` resolved against the FULL domain — a closed
    # set may be declared anywhere in the same chapter, not just the
    # current aggregate. `aggregates_by_name` is `domain_generator.rb`'s
    # own full-IR lookup (already built for `reference_checks`). `nil` —
    # not generated, not refused at runtime like Ruby's own
    # `undeclared_set` — when the target can't be resolved to an actual
    # closed-set value object at codegen time: there is no repository to
    # check against here, only a compile-time member table, so an
    # unresolvable target is the same "can't honestly generate this" shape
    # as every other skip in this generator.
    def admitted_set_members(admits, aggregates_by_name)
      aggregate_name, set_name = admits.to_s.split("::", 2)
      aggregate = aggregates_by_name[aggregate_name]
      return nil unless aggregate && set_name

      vo = aggregate[:value_objects].find { |v| v[:name] == set_name }
      return nil unless vo && vo[:closed_set]

      vo[:members].map { |row| row.first[1] }
    end

    # `InvariantViolation`/`admits_declared_set`. The template's own text
    # used to be reproduced here, byte-for-byte, as a baked `prefix`
    # (everything but the offered value, already substituted); V3 hands
    # the four declared arguments to the site's own typed `render_args`
    # instead, so the wording lives only in Vocabulary::RefusalTemplate
    # and the `.inspect`-quoting and ", " join of the member list are
    # `admitted`'s own RefusalSiteArgument row rather than this method's.
    # The member LIST is still resolved at codegen time
    # (`admitted_set_members`) and passed as a `&[&str]` literal; `{:?}`
    # on the offered `&str` produces the same quoted form Ruby's own
    # `.inspect` does for a plain string.
    def emit_admits_check(value_expr, attr, aggregates_by_name, value_objects_by_name)
      return nil unless attr[:admits]

      members = admitted_set_members(attr[:admits], aggregates_by_name)
      return nil unless members

      scalar, optional_source = optional_scalar_expr(value_expr, attr, value_objects_by_name)
      return nil unless scalar

      members_array = "[#{members.map(&:inspect).join(', ')}]"
      check = Exemplar.render(
        "admits_check",
        '["tmpl_member_a", "tmpl_member_b"]' => members_array,
        "tmpl_scalar" => scalar,
        '"tmpl_admits_name"' => attr[:name].to_s.inspect,
        '"tmpl_admits_target"' => attr[:admits].to_s.inspect
      )
      wrap_if_optional(check, optional_source)
    end

    # `TypeMismatch`/`pattern_mismatch`, the same way: four declared
    # arguments handed to the site's typed `render_args`, no copy of
    # `"{type}.{field} must match {pattern}, got {offered}"` left here.
    # `pattern` still goes over as the raw regex SOURCE string, not
    # re-escaped — Ruby's own template interpolates `attribute.pattern`
    # verbatim, and `pattern`'s own row says `quoting: "none"`.
    def emit_pattern_check(value_expr, attr, owner_type_name, value_objects_by_name)
      return nil unless attr[:pattern]

      scalar, optional_source = optional_scalar_expr(value_expr, attr, value_objects_by_name)
      return nil unless scalar

      check = Exemplar.render(
        "pattern_check",
        '"tmpl_pattern_text"' => attr[:pattern].inspect,
        "tmpl_scalar" => scalar,
        '"tmpl_pattern_owner"' => owner_type_name.to_s.inspect,
        '"tmpl_pattern_field"' => attr[:name].to_s.inspect
      )
      wrap_if_optional(check, optional_source)
    end
  end
end

File: rust/codegen/src/constraints.rs
//! Port of `rust/project/constraints.rb` — read that file's own header in
//! full; this mirrors `scalar_field_expr`/`admitted_set_members`/
//! `emit_admits_check`/`emit_pattern_check` directly.

use crate::exemplar::Exemplar;
use crate::json::Json;
use crate::naming;

/// The scalar `admitted_scalar`/`check_patterns` actually operate on.
pub fn scalar_field_expr(value_expr: &str, attr_type: &str, value_objects_by_name: &std::collections::HashMap<String, &Json>) -> Option<String> {
    if attr_type == "String" {
        return Some(value_expr.to_string());
    }
    let vo = value_objects_by_name.get(attr_type)?;
    let attrs = vo.get("attributes").map(Json::each).unwrap_or(&[]);
    if attrs.len() != 1 {
        return None;
    }
    let field = naming::rust_ident_field(crate::attr::name(&attrs[0]));
    Some(format!("{value_expr}.{field}"))
}

/// Mirrors `rust/project/constraints.rb`'s own `optional_scalar_expr`/
/// `wrap_if_optional` exactly — see that file's header comment for the
/// full reasoning (an OPTIONAL attribute is `Option<T>` in the struct,
/// not `T`; the two callers below used to hand `scalar_field_expr` the
/// bare field regardless, which compiles fine until `optional:` is
/// paired with `admits:`/`pattern:`). Returns the scalar expression to
/// check, and — only when the attribute is optional — the source
/// `Option` expression the caller must `if let Some(...)` around.
const OPTIONAL_BINDING: &str = "__optional_value";

fn optional_scalar_expr(
    value_expr: &str,
    attr: &Json,
    value_objects_by_name: &std::collections::HashMap<String, &Json>,
) -> Option<(String, Option<String>)> {
    if !crate::attr::optional(attr) {
        let scalar = scalar_field_expr(value_expr, crate::attr::type_name(attr), value_objects_by_name)?;
        return Some((scalar, None));
    }

    let scalar = scalar_field_expr(OPTIONAL_BINDING, crate::attr::type_name(attr), value_objects_by_name)?;
    Some((scalar, Some(value_expr.to_string())))
}

fn wrap_if_optional(check: String, optional_source: Option<String>) -> String {
    match optional_source {
        None => check,
        Some(source) => format!("if let Some({OPTIONAL_BINDING}) = &{source} {{ {check} }}"),
    }
}

/// `"Aggregate::SetName"` resolved against the full domain.
pub fn admitted_set_members(admits: &str, aggregates_by_name: &std::collections::HashMap<String, &Json>) -> Option<Vec<String>> {
    let mut parts = admits.splitn(2, "::");
    let aggregate_name = parts.next()?;
    let set_name = parts.next()?;
    let aggregate = aggregates_by_name.get(aggregate_name)?;
    let vos = aggregate.get("value_objects").map(Json::each).unwrap_or(&[]);
    let vo = vos.iter().find(|v| v.get("name").and_then(Json::as_str) == Some(set_name))?;
    if !vo.get("closed_set").map(Json::as_bool).unwrap_or(false) {
        return None;
    }
    let members = vo.get("members").map(Json::each).unwrap_or(&[]);
    Some(
        members
            .iter()
            .map(|row| {
                let pairs = row.as_array().unwrap_or(&[]);
                let first = pairs.first().and_then(|pair| pair.as_array()).unwrap_or(&[]);
                first.get(1).map(Json::to_s).unwrap_or_default()
            })
            .collect(),
    )
}

pub fn emit_admits_check(
    exemplar: &Exemplar,
    value_expr: &str,
    attr: &Json,
    aggregates_by_name: &std::collections::HashMap<String, &Json>,
    value_objects_by_name: &std::collections::HashMap<String, &Json>,
) -> Option<String> {
    let admits = crate::attr::admits(attr)?;
    let members = admitted_set_members(admits, aggregates_by_name)?;
    let (scalar, optional_source) = optional_scalar_expr(value_expr, attr, value_objects_by_name)?;

    let members_array = format!("[{}]", members.iter().map(|m| naming::ruby_inspect_string(m)).collect::<Vec<_>>().join(", "));
    let check = exemplar.render(
        "admits_check",
        &[
            ("[\"tmpl_member_a\", \"tmpl_member_b\"]", members_array),
            ("tmpl_scalar", scalar),
            ("\"tmpl_admits_name\"", naming::ruby_inspect_string(crate::attr::name(attr))),
            ("\"tmpl_admits_target\"", naming::ruby_inspect_string(admits)),
        ],
    );
    Some(wrap_if_optional(check, optional_source))
}

pub fn emit_pattern_check(exemplar: &Exemplar, value_expr: &str, attr: &Json, owner_type_name: &str, value_objects_by_name: &std::collections::HashMap<String, &Json>) -> Option<String> {
    let pattern = crate::attr::pattern(attr)?;
    let (scalar, optional_source) = optional_scalar_expr(value_expr, attr, value_objects_by_name)?;

    let check = exemplar.render(
        "pattern_check",
        &[
            ("\"tmpl_pattern_text\"", naming::ruby_inspect_string(pattern)),
            ("tmpl_scalar", scalar),
            ("\"tmpl_pattern_owner\"", naming::ruby_inspect_string(owner_type_name)),
            ("\"tmpl_pattern_field\"", naming::ruby_inspect_string(crate::attr::name(attr))),
        ],
    );
    Some(wrap_if_optional(check, optional_source))
}
